docker/build-env/Dockerfile

```diff
@@ -0,0 +1,137 @@
+FROM nvidia/cuda:10.1-devel-ubuntu18.04
+
+WORKDIR /root
+
+ENV DEBIAN_FRONTEND noninteractive
+ENV FORCE_UNSAFE_CONFIGURE 1
+
+# Install basic tools
+RUN apt-get update -qq && apt-get install -qq -y --no-install-recommends \
+    software-properties-common \
+    build-essential \
+    git tar wget curl && \
+    rm -rf /var/lib/apt/lists/*
+
+# Install cmake
+RUN wget -qO- "https://cmake.org/files/v3.17/cmake-3.17.0-Linux-x86_64.tar.gz" | tar --strip-components=1 -xz -C /usr/local
+
+# Install MPICH ABI compatible with Cray's lib on Piz Daint
+ARG MPICH_VERSION=3.1.4
+ARG MPICH_PATH=/usr/local/mpich
+RUN wget -q https://www.mpich.org/static/downloads/${MPICH_VERSION}/mpich-${MPICH_VERSION}.tar.gz && \
+    tar -xzf mpich-${MPICH_VERSION}.tar.gz && \
+    cd mpich-${MPICH_VERSION} && \
+    ./configure \
+      --disable-fortran \
+      --prefix=$MPICH_PATH && \
+    make install -j$(nproc) && \
+    rm -rf /root/mpich-${MPICH_VERSION}.tar.gz /root/mpich-${MPICH_VERSION}
+
+# Install MKL
+ARG MKL_VERSION=2020.0-088
+RUN wget -qO - https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB 2>/dev/null | apt-key add - && \
+    apt-add-repository 'deb https://apt.repos.intel.com/mkl all main' && \
+    apt-get install -y -qq --no-install-recommends intel-mkl-64bit-${MKL_VERSION} && \
+    rm -rf /var/lib/apt/lists/* && \ 
+    echo "/opt/intel/lib/intel64\n/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64" >> /etc/ld.so.conf.d/intel.conf && \
+    ldconfig
+
+# Install Boost
+ARG BOOST_MAJOR=1
+ARG BOOST_MINOR=72
+ARG BOOST_PATCH=0
+ARG BOOST_PATH=/usr/local/boost
+
+RUN wget -q https://dl.bintray.com/boostorg/release/${BOOST_MAJOR}.${BOOST_MINOR}.${BOOST_PATCH}/source/boost_${BOOST_MAJOR}_${BOOST_MINOR}_${BOOST_PATCH}.tar.gz -O boost.tar.gz && \
+    tar -xzf boost.tar.gz && \
+    cd boost_${BOOST_MAJOR}_${BOOST_MINOR}_${BOOST_PATCH} && \
+    ./bootstrap.sh --prefix=$BOOST_PATH && \
+    ./b2 -j$(nproc) && \
+    ./b2 install && \
+    rm -rf /root/boost.tar.gz /root/boost_${BOOST_MAJOR}_${BOOST_MINOR}_${BOOST_PATCH}
```

Comment:
I think we need only Boost headers.
@albestro Isn't HPX needing only them?

Reply:
ok, that's good to know. At any rate, the dependencies are not recompiled every time, but cached inside of the build environment docker image

Reply:
We depend on HPX 1.4.0
https://github.com/eth-cscs/DLA-Future/blob/4d92d14ba7e22bbcaa3475013626d9842f9572cc/CMakeLists.txt#L63

which [release notes](https://stellar-group.github.io/hpx/docs/sphinx/tags/1.4.0/html/releases/whats_new_1_4_0.html) states (I'm not really sure how to interpret it)

>We no longer require compiled Boost dependencies in certain configurations. This requires at least Boost 1.70, compiling on x86 with GCC 9, clang (libc++) 9, or VS2019 in C++17 mode.

In this issues you can track the status of the Boost dependencies in HPX https://github.com/STEllAR-GROUP/hpx/issues/3440

Reply:
HPX does not depend on any compiled Boost libraries anymore if 
- you are using Boost >= V1.70 (Boost.System is header only)
- you are using C++17 and a compiler that supports `std::filesystem`
- you are not interested in building and running the HPX inspect tool (depends on Boost.Regex)

We also reduce HPX's dependency on other Boost libraries as well, this however will take more time. The mentioned tickets documents our progress.